src/pkg/util/env/clean.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package env
+
+import (
+	"strings"
+
+	"github.com/opencontainers/runtime-tools/generate"
+	"github.com/singularityware/singularity/src/pkg/sylog"
+)
+
+// CleanEnv cleans environment variables before running the container
+func CleanEnv(g *generate.Generator, NoHome bool, IsCleanEnv bool, HomeDest string, environment []string) {
+	for _, env := range environment {
+		e := strings.SplitN(env, "=", 2)
+		if len(e) != 2 {
+			sylog.Verbosef("Can't process environment variable %s", env)
+			continue
+		}
+
+		// Transpose environment
+		if strings.HasPrefix(e[0], "SINGULARITYENV_") {
+			e[0] = strings.TrimPrefix(e[0], "SINGULARITYENV_")
+		} else if IsCleanEnv && (e[0] != "HOME" &&
+			e[0] != "TERM" &&
+			e[0] != "LANG" &&
+			e[0] != "http_proxy" &&
+			e[0] != "https_proxy" &&
+			e[0] != "no_proxy" &&
+			e[0] != "all_proxy" &&
+			e[0] != "ftp_proxy" &&
+			e[0] != "SINGULARITY_CONTAINER" &&
+			e[0] != "SINGULARITY_NAME") {
+			continue
+		}
+
+		if e[0] == "HOME" {
```

Comment:
This conditional doesn't make a lot of sense to me. I think we should always set `HOME=HomeDest` inside the container, whether or not the `HOME` variable is already set on this host. This conditional implies that we're only setting `HOME` in the container if `HOME` is *already* set on the host.

In my opinion, it makes much more sense to set the unconditional env variables outside of the loop where we iterate over the environment:

```
for ... {
    // iterating over env
}

if !NoHome {
    g.AddProcessEnv("HOME", HomeDest)
} else {
    g.AddProcessEnv("HOME", "/")
}

g.AddProcessEnv("LANG", "C")
```

Reply:
That's opposite current behavior:
```
$ /usr/local/singularity/2.6.0/bin/singularity shell -c --no-home ~/tmp/sing/centos7.img 
WARNING: Could not chdir to home: /home/jason
Singularity centos7.img:/> pwd
/
Singularity centos7.img:/> echo $HOME
/home/jason
```

What you want is when `--no-home` is specified, then `HOME=/`

Reply:
```
singularity exec --help 
[...] 
      --no-home            Do NOT mount users home directory if home is
                           not the current working directory.
```
so @jmstover is right, `--no-home` should just NOT mount HOME